PxGraf/Models/SavedQueries/ArchiveCube.cs

```diff
@@ -1,105 +1,87 @@
-﻿using Newtonsoft.Json;
-using PxGraf.Data;
-using PxGraf.Data.MetaData;
+﻿using Px.Utils.Language;
+using Px.Utils.Models;
+using Px.Utils.Models.Data.DataValue;
+using Px.Utils.Models.Metadata;
+using PxGraf.Models.Metadata;
+using PxGraf.Models.Queries;
+using PxGraf.Utility;
+using PxGraf.Utility.CustomJsonConverters;
 using System;
 using System.Collections.Generic;
+using System.Text.Json.Serialization;
 
 namespace PxGraf.Models.SavedQueries
 {
     /// <summary>
     /// Used for serializing DataCube objects to a file.
     /// </summary>
+    [JsonConverter(typeof(ArchiveMatrixSerializer))]
     public class ArchiveCube
     {
         /// <summary>
         /// When this archive object was originally created.
         /// </summary>
-        [JsonProperty("creationTime")]
-        public DateTime CreationTime { get; set; }
+        public DateTime CreationTime { get; }
 
         /// <summary>
         /// Contains all metadata of this cube.
         /// </summary>
-        [JsonProperty("meta")]
-        public CubeMeta Meta { get; set; }
+        public IReadOnlyMatrixMetadata Meta { get; }
 
         /// <summary>
         /// Collection of key (variable value code coordinates) value (double or missing code) pairs.
         /// </summary>
-        [JsonProperty("data")]
-        public List<double?> Data { get; set; }
+        [JsonConverter(typeof(DecimalDataValueListConverter))]
+        public List<DecimalDataValue> Data { get; }
 
         /// <summary>
         /// Notes mapped to the indexes of the data array.
         /// </summary>
-        [JsonProperty("dataNotes")]
-        public Dictionary<int, string> DataNotes { get; set; }
+        public Dictionary<int, MultilanguageString> DataNotes { get; }
 
         /// <summary>
-        /// Parameterles constructor for json deserialization.
+        /// Verison of the archive cube.
         /// </summary>
-        public ArchiveCube() { }
+        public string Version { get; }
 
-        private ArchiveCube(DataCube cube)
+        private ArchiveCube(IReadOnlyMatrixMetadata meta, IReadOnlyList<DecimalDataValue> data)
         {
             CreationTime = DateTime.Now;
+            Meta = meta;
 
-            Meta = cube.Meta.Clone();
-
-            List<double?> data = [];
-            Dictionary<int, string> dataNotes = [];
-
-            DataValue[] dataList = cube.Data;
-            int dataLen = dataList.Length;
-
-            for (int i = 0; i < dataLen; i++)
-            {
-                if (dataList[i].Type == DataValueType.Exist)
-                {
-                    data.Add(dataList[i].UnsafeValue);
-                }
-                else
-                {
-                    data.Add(null);
-                    dataNotes[i] = dataList[i].ToMachineReadableString(0);
-                }
-            }
-
-            Data = data;
-            DataNotes = dataNotes;
+            Data = [.. data];
```

Comment:
Could it be possible to use the same IReadOnlyList here, to avoid copying the contents of the data array